src/soep_preparation/dataset_merging/task.py

```diff
@@ -33,216 +58,147 @@ def _dataset_level(dataset: pd.DataFrame) -> str:
-        lambda left, right: merge_and_fillna_shared_cols(left, right),
-        datasets["household-time-varying"].values(),
+    time_constant_individual_data = merge_data_by_level(
+        name_to_df_mapping,
+        names_to_variables_by_level["individual-time-constant"],
     )
-    dataset_household_varying["hh_id_orig"] = dataset_household_varying.groupby(
-        "hh_id",
-    )["hh_id_orig"].ffill()
-    dataset_household_constant = reduce(
-        lambda left, right: merge_and_fillna_shared_cols(left, right),
-        datasets["household-time-constant"].values(),
+    time_varying_household_data = merge_data_by_level(
+        name_to_df_mapping,
+        names_to_variables_by_level["household-time-varying"],
     )
-    # Initialize with the first dataset
-    data_merged = dataset_individual_varying.reset_index(drop=True)
-
-    # Merge household-varying data
-    data_merged = merge_and_fillna_shared_cols(
-        data_merged,
-        dataset_household_varying,
-        merge_on=["hh_id", "survey_year"],
-        how="left",
+    time_constant_household_data = merge_data_by_level(
+        name_to_df_mapping,
+        names_to_variables_by_level["household-time-constant"],
     )
 
-    # Merge individual-constant data
-    data_merged = merge_and_fillna_shared_cols(
-        data_merged,
-        dataset_individual_constant,
-        merge_on=["hh_id_orig", "p_id"],
-        how="left",
+    # TODO (@felixschmitz)
+    # Make sure that there are time-invariant variables in the dataset
+    # Fill missing values (due to no observations in the survey year) with pd.NA
+    return merge_data(
+        [
+            time_varying_individual_data,
+            time_constant_individual_data,
+            time_varying_household_data,
+            time_constant_household_data,
+        ],
     )
-    # Merge household-constant data
-    data_merged = merge_and_fillna_shared_cols(
-        data_merged,
-        dataset_household_constant,
-        merge_on=["hh_id"],
-        how="left",
-    )
-
-    # Set the final index
-    data_merged = data_merged.set_index(["p_id", "survey_year"]).sort_index()
-
-    # Merge duplicate columns originating from distinct datasets
-    return combine_similar_information_from_different_columns(data_merged)
 
 
-DATASETS = _datasets_to_merge(DATA_CATALOGS)
-
-
-def task_merge_datasets(
-    datasets: Annotated[dict, DATASETS],
-) -> Annotated[pd.DataFrame, DATA_CATALOGS["merged"]["merged_dataset"]]:
-    """Merge initially on observational and time-varying level, then aggregate datasets.
-
-    Args:
-        datasets (dict): Dictionary containing the datasets to be merged.
-
-    Returns:
-        pd.DataFrame: Merged panel dataset of cleaned SOEP datasets.
-    """
-    datasets_by_level = structure_datasets_by_level(datasets)
-    datasets_with_origin_dummy = _add_origin_dummy(datasets_by_level)
-    return merge_datasets(datasets_with_origin_dummy)
+def task_access_variables():
```

Comment:
Extract this function to a separate module `task_example.py` and save the rest of the module under a name that does not get caught by pytask. Good to have it as a "test", but that's it.

Project-level code will run the pipeline and then set up their own task(s) based on the `create_...` function.